### PR TITLE
Support for GitHub Actions as testing service

### PR DESCRIPTION
**Pull request by ilveroluca**
Opening this PR to collect the implementation of support for GitHub Actions.  When complete will solve issue #104

### CODE REVIEW

File: tests/config/data/make-test-rocrates.py [hunk 1]
```diff
@@ -134,10 +139,26 @@ def patch_metadata_graph_node(metadata_file, node, properties):
                               'name': 'sort-and-change-case-invalid-service-type'
                           })
 
+# GitHub Actions crate
+patch_metadata_graph_node('crates/ro-crate-galaxy-sortchangecase-github-actions/ro-crate-metadata.json',
+                          node=("@type", "TestInstance"),
+                          properties={
+                              'url': 'https://api.github.com',
+                              'resource': '/repos/crs4/life_monitor/actions/workflows/4094661',
```
Comment (simleo) [inline tests/config/data/make-test-rocrates.py:147]: `resource` is a relative path in Jenkins and Travis instances. We need to stick to the same convention here.

File: lifemonitor/api/models/services/github.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        # obvious way to istantiate a PyGithub WorkflowRun object given a build
+        # number -- but there's has to be a way.  We can easily asseble the URL
+        # of the request to directly retrive the data we need here.
+        assert isinstance(build_number, int)
+        for run in self._iter_runs(test_instance):
+            if run.id == build_number:
+                return GithubTestBuild(self, test_instance, run)
+        raise lm_exceptions.EntityNotFoundException(models.TestBuild, entity_id=build_number)
+
+    def _parse_workflow_url(self, resource: str) -> Tuple[str, str, str]:
```
Comment (simleo) [inline lifemonitor/api/models/services/github.py:156]: Does not use `self`. Could be changed to a static method or a plain function outside of the class.

File: lifemonitor/api/models/services/github.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+        """
+        Given a URL to the testing Github Worklflow, returns a tuple
+        (server, repository, workflow_id)
+        """
+        # URL identifies a Github Workflow and should have the form:
+        #   https://api.github.com/repos/crs4/life_monitor/actions/workflows/4094661
+        expected_url_msg = " Expected '/repos/{org}/{reponame}/actions/workflows/{workflow_id}'"
```
Comment (simleo) [inline lifemonitor/api/models/services/github.py:163]: Error msg does not match the comment. `https://api.github.com/repos/crs4/life_monitor/actions/workflows/4094661` is accepted, so the comment seems to be right.

File: lifemonitor/api/models/services/github.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        try:
+            logger.debug("Parsing testing instance resource URL '%s'", resource)
+            result = urlparse(resource)
+            server = f"{result.scheme}://{result.netloc}"
+            # simply split the path using the / delimiter.  We'll later access the parts by index
+            parts = result.path.split('/')
+            # Validate our assumptions
+            if len(parts) != 7:
```
Comment (simleo) [inline lifemonitor/api/models/services/github.py:171]: Change the parsing logic to accept a relative path, as in the travis and jenkins services. Actually, we should accept both for robustness (in the end, the resource is always joined to a base URI).

File: tests/config/data/make-test-rocrates.py (hunk 1 above)
Comment (simleo) [reply tests/config/data/make-test-rocrates.py]: Done in ac54584

File: lifemonitor/api/models/services/github.py (hunk 3 above)
Comment (simleo) [reply lifemonitor/api/models/services/github.py]: Done in ac54584. *Also* accepts a path with no leading slash now.

File: lifemonitor/api/models/services/github.py (hunk 4 above)
Comment (simleo) [reply lifemonitor/api/models/services/github.py]: Done in ac54584.

**Review by kikkomep — APPROVED** [review]
LGTM

File: lifemonitor/api/models/services/github.py [hunk 5]
```diff
@@ -38,6 +39,9 @@
 logger = logging.getLogger()
 
 
+RESOURCE_PATTERN = re.compile(r"/?repos/(?P<org>[^/]+)/(?P<repo>[^/]+)/actions/workflows/(?P<wf>[^/]+)")
```
[the remark this answers is not in the document]
Comment (simleo) [reply lifemonitor/api/models/services/github.py]: done

File: lifemonitor/api/models/services/github.py [hunk 6]
```diff
@@ -160,23 +164,14 @@ def _parse_workflow_url(self, resource: str) -> Tuple[str, str, str]:
...
-            if parts[1] != 'repos':
-                raise RuntimeError("Missing or misplaced /repos in URL to Github Workflow." + expected_url_msg)
-            if parts[4] != 'actions' or parts[5] != 'workflows':
-                raise RuntimeError("Missing or misplaced /actions/workflows in URL to Github Workflow." + expected_url_msg)
-            # Extract info
-            repository = '/'.join(parts[2:4])
-            workflow_id = parts[6]
+            m = RESOURCE_PATTERN.match(result.path)
+            if not m:
+                raise RuntimeError("Malformed GitHub workflow path. Expected: 'repos/{org}/{reponame}/actions/workflows/{workflow_id}'")
```
[the remark this answers is not in the document]
Comment (simleo) [reply lifemonitor/api/models/services/github.py]: done

File: lifemonitor/api/models/services/github.py (hunk 2 above)
Comment (simleo) [reply lifemonitor/api/models/services/github.py]: done

**Pull request merged**